dbms/src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -1186,22 +1187,74 @@ static size_t getLimitForSorting(ASTSelectQuery & query)
 void InterpreterSelectQuery::executeOrder(Pipeline & pipeline)
 {
     SortDescription order_descr = getSortDescription(query);
-    size_t limit = getLimitForSorting(query);
+
+    if (order_descr.empty())
+        return;
 
     const Settings & settings = context.getSettingsRef();
+    size_t limit = getLimitForSorting(query);
 
-    pipeline.transform([&](auto & stream)
+    bool use_sorting = true;
+
+    auto order = order_descr[0].direction;
+
+    if (auto merge_tree = dynamic_cast<StorageMergeTree *>(storage.get()))
     {
-        auto sorting_stream = std::make_shared<PartialSortingBlockInputStream>(stream, order_descr, limit);
+        if (!query.distinct && !query.group_expression_list)
+        {
+            auto aliases = query_analyzer->getAliases();
+            merge_tree->do_not_read_with_order = true;
+            auto column_sorted_order = merge_tree->getData().getSortColumns();
 
-        /// Limits on sorting
-        IProfilingBlockInputStream::LocalLimits limits;
-        limits.mode = IProfilingBlockInputStream::LIMITS_TOTAL;
-        limits.size_limits = SizeLimits(settings.max_rows_to_sort, settings.max_bytes_to_sort, settings.sort_overflow_mode);
-        sorting_stream->setLimits(limits);
+            for (size_t i = 0; i < order_descr.size(); ++i)
+            {
+                if ((i == column_sorted_order.size())
+                    || (order_descr[i].column_name != column_sorted_order[i])
+                    || (order != order_descr[i].direction)
+                    || aliases.find(order_descr[i].column_name) != aliases.end())
+                    break;
 
-        stream = sorting_stream;
-    });
+                if (i == order_descr.size() - 1)
+                {
+                    /// Threads can not steal task. (for order)
+                    merge_tree->do_not_read_with_order = false;
+                    use_sorting = false;
+
+                    pipeline.transform([&](auto & stream)
+                    {
+                        auto async_stream = std::make_shared<AsynchronousBlockInputStream>(stream);
+                        stream = async_stream;
+                    });
+
+                    if (order == -1)
+                    {
+                        pipeline.transform([&](auto & stream)
+                        {
+                            auto reverse_stream = std::make_shared<ReverseBlockInputStream>(stream);
+                            stream = reverse_stream;
```

Comment:
You can write it in single statement to avoid excessive operations with refcounts.